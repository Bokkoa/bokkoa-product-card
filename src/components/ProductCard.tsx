import React, { CSSProperties, createContext } from 'react'
import styles from '../styles/styles.module.css'
import { useProduct } from '../hooks/useProduct'
import { InitialValues, Product, ProductCardHandlers, ProductContextProps, onChangeArgs } from '../interfaces/interfaces'


export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext


export interface ProductCardProps {
  product: Product
  // children?: ReactElement | ReactElement[]
  children: (args: ProductCardHandlers) => JSX.Element
  className?: string
  style?: CSSProperties
  onChange?: ( args: onChangeArgs ) => void;
  value?: number;
  initialValues?: InitialValues
}


export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: ProductCardProps) => {

  const { counter, increaseBy, maxCount, reset, isMaxCountReached } = useProduct({ onChange, product, value, initialValues })

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount
      }}
    >
      <div 
        className={`${styles.productCard} ${className}`}
        style={style}
      >
        {/* STATE INITIALIZER AND FUNCTION CHILD PATTERN IMPLEMENTED */}
        {children({
          count: counter,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          isMaxCountReached,
          reset
        })}
      </div>
    </Provider>
  )
}
