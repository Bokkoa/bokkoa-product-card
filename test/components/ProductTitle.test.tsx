import * as React from 'react';
import renderer from 'react-test-renderer'
import ProductCard, { ProductTitle } from '../../src/components'
import { product1 } from '../data/products';

describe('ProductTitle', () => { 

  test('should show the component with the custom title correctly', () => { 
    const wrapper = renderer.create(
      <ProductTitle title='Custom Product' />
    )

    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  test('should show the component with product\'s name', () => { 
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {
          () => (
            <ProductTitle title='Custom Product' />
          )
        }
      </ProductCard>
    )
    expect(wrapper.toJSON()).toMatchSnapshot()

  })
})