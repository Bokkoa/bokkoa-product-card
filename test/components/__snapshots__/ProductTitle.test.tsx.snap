// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should show the component with product's name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Custom Product
  </span>
</div>
`;

exports[`ProductTitle should show the component with the custom title correctly 1`] = `
<span
  className="productDescription undefined"
>
  Custom Product
</span>
`;
